File: duyo-backend/src/duyo/api/v1/dtm.py
"""DTM practice questions — AI-generated, grounded in the textbook RAG."""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from duyo.api.deps import get_current_user, get_db
from duyo.models.user import User
from duyo.schemas.dtm import DtmQuestion, DtmRequest, DtmResponse
from duyo.services.dtm import generate_questions

router = APIRouter(prefix="/dtm", tags=["dtm"])


@router.post("/questions", response_model=DtmResponse)
async def dtm_questions(
    payload: DtmRequest,
    _user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> DtmResponse:
    """Generate grounded DTM-style multiple-choice questions for a subject."""
    qs = await generate_questions(db, subject=payload.subject, count=payload.count)
    return DtmResponse(questions=[DtmQuestion(**q) for q in qs])

File: duyo-mobile/src/api/endpoints/dtm.ts
import { AI_TIMEOUT_MS, apiClient } from '@/api/client';
import type { DTMSubject } from '@/mocks/dtm';

// Mirrors backend duyo.schemas.dtm. Unlike /puzzles, the DTM endpoint ships
// `correct_index` inside the question payload — the backend has no "check my
// answer" route for these, because every set is generated fresh and never
// persisted. So the grading below is done against the server's own answer key,
// not against anything the client made up.

export interface DtmQuestion {
  text: string;
  choices: string[];
  correct_index: number;
  explanation: string;
}

/** How many questions one practice run asks for. Backend caps at 10. */
export const DTM_QUESTION_COUNT = 5;

/**
 * The subject slug the backend indexes textbooks under
 * (duyo.textbook.pipeline._SUBJECT_ALIASES). The screen's own keys are UI
 * labels shared with lesson-help; these are the RAG corpus names, and sending
 * the wrong one silently retrieves nothing.
 */
const SUBJECT_SLUG: Record<DTMSubject, string> = {
  math: 'matematika',
  physics: 'fizika',
  chemistry: 'kimyo',
  native: 'ona-tili',
  history: 'tarix',
};

/**
 * Generate a fresh set of DTM-style questions for a subject.
 *
 * The backend grounds these in the ingested textbook chunks for that subject.
 * It answers 200 with an EMPTY list — never an error — when it has nothing to
 * ground on or the model call fails, so an empty result is a state the caller
 * must render honestly rather than a bug to swallow.
 */
export async function fetchDtmQuestions(
  subject: DTMSubject,
  count: number = DTM_QUESTION_COUNT,
): Promise<DtmQuestion[]> {
  const { data } = await apiClient.post<{ questions: DtmQuestion[] }>(
    '/dtm/questions',
    { subject: SUBJECT_SLUG[subject], count },
    // Generation, not CRUD — the default 15s cut this off mid-thought.
    { timeout: AI_TIMEOUT_MS },
  );
  // Defend the quiz loop: a malformed row would render a question with no way
  // to answer it right. Dropping it is honest — the child sees fewer questions,
  // never a broken one.
  return (data.questions ?? []).filter(
    (q) =>
      !!q.text &&
      Array.isArray(q.choices) &&
      q.choices.length >= 2 &&
      q.correct_index >= 0 &&
      q.correct_index < q.choices.length,
  );
}
